oak_functions/wasm/src/lib.rs

```diff
@@ -36,388 +35,452 @@ use oak_functions_abi::{
+            ctx.as_context_mut(),
+            &[wasmi::Value::I32(len)],
+            &mut address,
+        )
+        .expect("`alloc` call failed");
+    let dest_ptr = match address[0] {
+        // Assumes 32-bit Wasm addresses.
+        wasmi::Value::I32(v) => v
+            .try_into()
+            .map_or_else(|_| Err(OakStatus::ErrInternal), Ok),
+        _ => Err(OakStatus::ErrInternal),
+    }?;
+
+    // Write to buffer.
+    write_buffer(ctx, memory, &buffer, dest_ptr)?;
+    write_u32(ctx, memory, dest_ptr, dest_ptr_ptr)?;
+    write_u32(ctx, memory, buffer.len() as u32, dest_len_ptr)?;
+    Ok(())
+}
+
+/// Corresponds to the host ABI function [`read_request`](https://github.com/project-oak/oak/blob/main/docs/oak_functions_abi.md#read_request).
+pub fn read_request<L: OakLogger>(
+    caller: &mut wasmi::Caller<'_, UserState<L>>,
+    dest_ptr_ptr: AbiPointer,
+    dest_len_ptr: AbiPointer,
+) -> Result<(), OakStatus> {
+    let alloc = get_exported_alloc(caller)?;
+    let mut memory = get_exported_memory(caller)?;
+    let request_bytes = caller.data().request_bytes.clone();
+
+    alloc_and_write_buffer(
+        caller,
+        &mut memory,
+        alloc,
+        request_bytes,
+        dest_ptr_ptr,
+        dest_len_ptr,
+    )
+}
+
+/// Corresponds to the host ABI function [`write_response`](https://github.com/project-oak/oak/blob/main/docs/oak_functions_abi.md#write_response).
+pub fn write_response<L: OakLogger>(
+    caller: &mut wasmi::Caller<'_, UserState<L>>,
+    buf_ptr: AbiPointer,
+    buf_len: AbiPointerOffset,
+) -> Result<(), OakStatus> {
+    let mut memory = get_exported_memory(caller)?;
+    caller.data_mut().response_bytes = read_buffer(caller, &mut memory, buf_ptr, buf_len)?;
+    Ok(())
+}
+
+pub fn invoke_extension<L: OakLogger>(
+    caller: &mut wasmi::Caller<'_, UserState<L>>,
+    handle: i32,
+    request_ptr: AbiPointer,
+    request_len: AbiPointerOffset,
+    response_ptr_ptr: AbiPointer,
+    response_len_ptr: AbiPointer,
+) -> Result<(), OakStatus> {
+    let alloc = get_exported_alloc(caller)?;
+
+    let mut memory = get_exported_memory(caller)?;
+
+    let request = read_buffer(caller, &mut memory, request_ptr, request_len).map_err(|err| {
+        caller.data().log_error(&format!(
+            "Handle {:?}: Unable to read input from guest memory: {:?}",
+            handle, err
+        ));
+        OakStatus::ErrInvalidArgs
+    })?;
+
+    let user_state = caller.data_mut();
+    let extension = user_state.get_extension(handle)?;
+    let response = extension.invoke(request)?;
+
+    alloc_and_write_buffer(
+        caller,
+        &mut memory,
+        alloc,
+        response,
+        response_ptr_ptr,
+        response_len_ptr,
+    )
+}
+
+// Helper function to get exported MEMORY from caller.
+fn get_exported_memory<L: OakLogger>(
+    caller: &mut wasmi::Caller<'_, UserState<L>>,
+) -> Result<Memory, OakStatus> {
+    let ext = caller.get_export(MEMORY_NAME).ok_or({
```

Comment:
May be more idiomatic to use https://doc.rust-lang.org/std/result/enum.Result.html#method.inspect_err for the logging, and then do the conversion only in ok_or, but up to you.

Reply:
+1. I *think* the current implementation will always log, even if there are no errors, since the value would always be calculated but only returned if there was an error. The other option is to use ok_or_else and do the logging in the closure. In that case it will only call the closure if there is an error.

Reply:
Ah yes I didn't even realise that, but it's true; here the value will always be evaluated, which looks like it will always print an error, regardless of the actual value? But that should be easy to falsify by looking at logs.